### ISSUE TITLE
Feature Request: Primer on C / Chapel interoperability

### ISSUE THREAD

**Issue by buddha314** [chapel-lang/chapel#11858]
I would find it very useful to see a complete example of calling Chapel modules / files / functions from C and linked from the [Primers](https://chapel-lang.org/docs/latest/primers/index.html) page.

There are a few things of particular interest
1. Calling Chapel from C
1. Passing matrices / arrays (sparse or dense) back and forth, perhaps with `boost` vectors on the C end.
1. Importing full Chapel modules
1. How to link to Chapel using `CMake` (maybe a `find_chapel()` macro?)
1. Perhaps a github repo with a reference directory structure

**Comment by lydia-duncan** [comment]
I think this is a reasonable request.  I probably won't get to it until closer to feature freeze, but definitely will before next release

**Comment by buddha314** [comment]
> 
> 
> Just to be sure, do you already know about https://chapel-lang.org/docs/technotes/libraries.html ?

Yes, I do, thanks for checking in!  It seems much of this information is spread around and would be useufl to a crappy programmer like me to see it laid out in one place, especially on the matrix / vec ops.

"How do I always re-appear right before code freeze?" You ask. Talent, skill, dedication... :)

**Comment by lydia-duncan** [comment]
At least this time you are three months ahead of it! XD

**Comment by lydia-duncan** [comment]
I have a start of this over at [this branch](https://github.com/chapel-lang/chapel/compare/master...lydia-duncan:interopPrimer)

Things that aren't covered yet:
- [ ] extern records
   - [ ] with omitted fields
- [ ] renaming extern symbols using the "external name" syntax
- [ ] comments describing how to include the code that defines the extern symbols
  - [ ] via `require` statements
  - [ ] via the compile line command
- [ ] types that are already available without the user having to write `extern type` declarations (i.e. c_int, etc)
- [ ] passing arrays back and forth
  - [ ] c_ptr
  - [ ] c_array
  - [ ] chpl_external_array
  - [ ] chpl_opaque_array
- [ ] update primers/README
- [ ] update primers/Makefile 
- [ ] something similar for Python
- [ ] something similar for Fortran

**Comment by stonea** [comment]
On https://github.com/Cray/chapel-private/issues/2126 I wrote some notes down on what I found confusing in the existing interop documentation. Regarding calling a C function from Chapel: one thing I’d advocate for in the primer is explicit instructions on how to invoke the “chpl” command (passing a header and c file to it). It seems obvious in retrospect, but it took me a while to figure out that I needed to explicitly pass the .h file (in addition to having the implementation in a .c file).

**Comment by bradcray** [comment]
FWIW (re-reading Lydia's [comment above](https://github.com/chapel-lang/chapel/issues/11858#issuecomment-472067955)), this primer need not do everything in the first draft.  Even a very simple/minimal interop primer would be better than none.

One minor note on Andy's comment just to make sure there's no misunderstanding: Note that external C routines can also be satisfied via `.o`, `.a`, `.so` files as well as `.c`.

**Comment by lydia-duncan** [comment]
> FWIW (re-reading Lydia's [comment above](https://github.com/chapel-lang/chapel/issues/11858#issuecomment-472067955)), this primer need not do everything in the first draft. Even a very simple/minimal interop primer would be better than none.

For sure!  I've just been bogged down with other tasks by the time doc/code freeze has rolled around and haven't been able to prioritize it at other points.

**Comment by bradcray** [comment]
Not calling you out specifically, just in case others felt motivated to pick this up and were intimidated by the list of checkboxes.
